pkg/resources/address/address.go

```diff
@@ -207,35 +207,46 @@ func (m *ModifiersBuilder) Build(ctx context.Context) ([]func(*kubermaticv1.Clus
 
 func (m *ModifiersBuilder) getFrontProxyLBServiceData(frontProxyLoadBalancerService *corev1.Service) (string, string) {
 	//  frontProxyLBServiceIP is set according to below priority
-	// 1. First public IP from the status list
-	// 2. First private IP from the status list
-	// 3. Default IP as per configured spec if status is not populated
+	// 1. First public IPv4 from the status list
+	// 2. First private IPv4 from the status list
+	// 3. First public IPv6 from the status list
+	// 4. First private IPv6 from the status list
+	// 5. Default IP as per configured spec if status is not populated
 	serviceIP := ""
 	serviceHostname := ""
-	if len(frontProxyLoadBalancerService.Status.LoadBalancer.Ingress) > 0 {
-		var tmpIP string
-		for _, ingress := range frontProxyLoadBalancerService.Status.LoadBalancer.Ingress {
-			if ingress.IP != "" && !net.ParseIP(ingress.IP).IsPrivate() && !utilnet.IsIPv6String(ingress.IP) && tmpIP == "" {
-				tmpIP = ingress.IP
-			}
-			if ingress.Hostname != "" {
-				serviceHostname = ingress.Hostname
-			}
-		}
+	var publicIPv4, privateIPv4, privateIPv6, publicIPv6 []string
 
-		if tmpIP != "" {
-			serviceIP = tmpIP
-		} else {
-			// select first non-ipv6 private IP
-			for _, ingress := range frontProxyLoadBalancerService.Status.LoadBalancer.Ingress {
-				if !utilnet.IsIPv6String(ingress.IP) {
-					serviceIP = ingress.IP
-					break
-				}
-			}
+	for _, ingress := range frontProxyLoadBalancerService.Status.LoadBalancer.Ingress {
+		if ingress.IP != "" && !net.ParseIP(ingress.IP).IsPrivate() && !utilnet.IsIPv6String(ingress.IP) {
+			publicIPv4 = append(publicIPv4, ingress.IP)
+		}
+		if ingress.IP != "" && net.ParseIP(ingress.IP).IsPrivate() && !utilnet.IsIPv6String(ingress.IP) {
+			privateIPv4 = append(privateIPv4, ingress.IP)
+		}
+		if ingress.IP != "" && utilnet.IsIPv6String(ingress.IP) && !net.ParseIP(ingress.IP).IsPrivate() {
+			publicIPv6 = append(publicIPv6, ingress.IP)
+		}
+		if ingress.IP != "" && utilnet.IsIPv6String(ingress.IP) && net.ParseIP(ingress.IP).IsPrivate() {
+			privateIPv6 = append(privateIPv6, ingress.IP)
+		}
```

Comment:
Do we need to do `ingress.IP != ""` and `net.ParseIP(ingress.IP).IsPrivate()` really 4 times? And `utilnet.IsIPv6String(ingress.IP)` twice?

Reply:
Refactored it a bit